torch/_higher_order_ops/hints_wrap.py

```diff
@@ -0,0 +1,105 @@
+# mypy: allow-untyped-defs
+import torch
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    autograd_not_implemented,
+    reenter_make_fx,
+    unique_graph_id,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+
+
+# used for wrapping a function/op with context hints
+class HintsWrapper(HigherOrderOperator):
+    def __init__(self):
+        super().__init__("hints_wrapper")
+
+    def __call__(self, body_fn, args, kwargs, hints):
+        if not isinstance(args, tuple):
+            raise RuntimeError(f"args must be a tuple, got {type(args)}")
+
+        if not all(isinstance(t, (torch.Tensor, int, float, bool)) for t in args):
+            raise RuntimeError(
+                "args must be a tuple of tensors, ints, floats, or bools, got "
+                f"{args}"
+            )
+
+        if not isinstance(kwargs, dict):
+            raise RuntimeError(f"kwargs must be a dict, got {type(kwargs)}")
+
+        if not isinstance(hints, dict):
```

Comment:
Do we want to have more type checking for the contents of hints (like python primitive types, str, int, float, bool, maybe some pytree structure) or should we allow everything to be possible in hints, which seems a bit risky.

Reply:
Exactly, IMO it's not necessary to support arbitrary types, meanwhile as you said it's not safe. I can add checks for native python types like str, int, float, bool. For pytree, is it possible to  support in another PR?

Reply:
Sounds good. I think it's more important to support pytree inputs/outputs (pytree value hints is not that important). We can do it next pr probably. Something like below:
```python
def f(x):
  return [x[0].sin() + x[1]["val"].cos(), x[0].cos()[
hints_wrapper(f, ((t1, {"val": t2}))...)
```

If we go through torch.compile, this is supported (dynamo can flatten the inputs and outputs) but if we're not going through torch.compile, we probably can error out early (probably this is what we should do in this pr) or if we really want to support it, we could borrow some inspiration from map_wrapper.

Reply:
If hints_wrapper with pytree inputs runs into eager, I think the check at Line24 will fail.